mixer/pkg/runtime2/config/ephemeral.go

```diff
@@ -0,0 +1,370 @@
+	handlers := make(map[string]*Handler)
+
+	for key, resource := range e.entries {
+		var info *adapter.Info
+		var found bool
+		if info, found = e.adapters[key.Kind]; !found {
+			continue
+		}
+
+		adapterName := key.String()
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming handler config: name='%s'\n%s", adapterName, resource.Spec.String())
+		}
+
+		cfg := &Handler{
+			Name:    adapterName,
+			Adapter: info,
+			Params:  resource.Spec,
+		}
+
+		handlers[cfg.Name] = cfg
+	}
+
+	counters.handlerConfig.Add(float64(len(handlers)))
+	return handlers
+}
+
+func (e *Ephemeral) processInstanceConfigs(counters Counters) map[string]*Instance {
+	instances := make(map[string]*Instance)
+
+	for key, resource := range e.entries {
+		var info *template.Info
+		var found bool
+		if info, found = e.templates[key.Kind]; !found {
+			continue
+		}
+
+		instanceName := key.String()
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming instance config: name='%s'\n%s", instanceName, resource.Spec.String())
+		}
+
+		cfg := &Instance{
+			Name:     instanceName,
+			Template: info,
+			Params:   resource.Spec,
+		}
+
+		instances[cfg.Name] = cfg
+	}
+
+	counters.instanceConfig.Add(float64(len(instances)))
+	return instances
+}
+
+func (e *Ephemeral) processRuleConfigs(
+	handlers map[string]*Handler,
+	instances map[string]*Instance,
+	counters Counters) []*Rule {
+
+	log.Debug("Begin processing rule configurations.")
+
+	var rules []*Rule
+
+	for ruleKey, resource := range e.entries {
+		if ruleKey.Kind != RulesKind {
+			continue
+		}
+		counters.ruleConfig.Add(1)
+
+		ruleName := ruleKey.String()
+
+		cfg := resource.Spec.(*config.Rule)
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming rule: name='%s'\n%s", ruleName, cfg.String())
+		}
+
+		// TODO(Issue #2139): resourceType is used for backwards compatibility with labels: [istio-protocol: tcp]
+		// Once that issue is resolved, the following block should be removed.
+		rt := resourceType(resource.Metadata.Labels)
+		if cfg.Match != "" {
+			m, err := expr.ExtractEQMatches(cfg.Match)
+			if err != nil {
+				log.Warnf("ConfigWarning: Unable to extract resource type from rule: name='%s'", ruleName)
+
+				// instead of skipping the rule, add it to the list. This ensures that the behavior will
+				// stay the same when this block is removed.
+			} else {
+				if ContextProtocolTCP == m[ContextProtocolAttributeName] {
+					rt.protocol = protocolTCP
+				}
+			}
+		}
+
+		// extract the set of actions from the rule, and the handlers they reference.
+		actions := make([]*Action, 0, len(cfg.Actions))
+		for i, a := range cfg.Actions {
+			log.Debugf("Processing action: %s[%d]", ruleName, i)
+
+			var found bool
+			var handler *Handler
+			handlerName := canonicalize(a.Handler, ruleKey.Namespace)
+			if handler, found = handlers[handlerName]; !found {
+				log.Warnf("handler not found: handler='%s', action='%s[%d]'",
```

Comment:
What does it mean that the handler is not found? Is this likely due to the eventually consistent nature of the config update process? Is it reasonable to output a warning in the log for that case then? Seems like this would alarm the operator for something that is not actionable, has no practical effect, and will get fixed automatically in a second or two.

Reply:
Let's assume that static checking is in place.

In that case, config mismatches are a result of eventual consistency inconsistencies, or due to a crash while configs were in the process of being pushed.

Inconsistencies are inherently transient and will resolve themselves within a minute. So for those cases we shouldn't produce any warnings in the log, since the system is running fine. This case will be extremely rare in the real world since config is distributed in typically much less time than 1 minute.

The other case, where config is hopelessly confused to a partial push, that's a real error condition. We want to warn about this in some way. This case should also be very rare.

Given the fact both conditions are very rare, and our inability to tell the difference between the two cases, then I suppose it doesn't hurt to always log. Should these be warnings or errors however? The inconsistency case should be a warning, the busted config should be an error. But again since we can't tell the difference between the two cases, I guess we have to pick one. I'd go with error rather than warning.

Reply:
Error, as in, log as Error but otherwise go ahead and process the config, or, as in, refuse the config? The latter is not easy to do meaningfully in the current config model.

Reply:
It will always be possible to bypass the static checks, therefore runtime *should* inform user that something is not right. Eventual consistency may not be rate in large clusters, so logging every time may be spammy. 

At the cost of more complexity, we can be very precise. Remember the first rejection and only warn on it after certain time, say a 30s/1min. That way users **will not** see eventual consistency errors.